ejb3/src/main/java/org/jboss/as/ejb3/component/stateful/StatefulSessionComponentInstance.java

```diff
@@ -72,6 +75,9 @@
      */
     private boolean synchronizationRegistered = false;
 
+    private final AtomicReference<Boolean> delayedAfterCompletion = new AtomicReference<>(false);
```

Comment:
I think you probably want to combine these atomics into one combined state variable that can be atomically queried and set.